src/liballoc/sync.rs

```diff
@@ -751,6 +751,76 @@ impl<T: ?Sized> Arc<T> {
         this.inner().strong.load(SeqCst)
     }
 
+    /// Increments the strong reference count on the `Arc<T>` associated with the
+    /// provided pointer by one.
+    ///
+    /// # Safety
+    ///
+    /// The pointer must have been obtained through `Arc::into_raw`, and the
+    /// associated `Arc` instance must be valid (i.e. the strong count must be at
+    /// least 1) for the duration of this method.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(arc_mutate_strong_count)]
+    ///
+    /// use std::sync::Arc;
+    ///
+    /// let five = Arc::new(5);
+    ///
+    /// unsafe {
+    ///     let ptr = Arc::into_raw(five);
+    ///     Arc::incr_strong_count(ptr);
+    ///
+    ///     // This assertion is deterministic because we haven't shared
+    ///     // the `Arc` between threads.
+    ///     let five = Arc::from_raw(ptr);
+    ///     assert_eq!(2, Arc::strong_count(&five));
+    /// }
+    /// ```
+    #[inline]
+    #[unstable(feature = "arc_mutate_strong_count", issue = "none")]
+    pub unsafe fn incr_strong_count(ptr: *const T) {
+        let arc = Arc::from_raw(ptr);
```

Comment:
I would recomment to simply copy the implementation from here: https://github.com/rust-lang/futures-rs/blob/4fef913048cc083f824a5cc4d22d2beb435bb936/futures-task/src/waker.rs#L34-L38

This one is also correct in case `Arc::clone` panics, while the implementation here is not (it will `drop(arc)`, which leads to a a decremented refcount).

Reply:
Ah that's amazing, thank you.